<launch>
  <!-- Topics -->
  <param name="/use_sim_time" value="true" />

  <arg name="input_sensor_points_topic" default="/filtered_points" doc="Sensor points topic"/>
  <arg name="input_initial_pose_topic" default="/ekf_pose_with_covariance" doc="Initial position topic to align" />
  <arg name="input_map_points_topic" default="/points_map" doc="Map points topic" />
  <arg name="output_pose_topic" default="ndt_pose" doc="Estimated self position" />
  <arg name="output_pose_with_covariance_topic" default="ndt_pose_with_covariance" doc="Estimated self position with covariance" />
  <arg name="output_diagnostics_topic" default="diagnostics" doc="Diagnostic topic" />

  <include file="$(find ndt_localizer)/launch/static_tf.launch" />
  <include file="$(find ndt_localizer)/launch/map_loader.launch" />
  <include file="$(find ndt_localizer)/launch/points_downsample.launch" />

  <!-- tf -->
  <!-- <arg name="base_frame" default="base_link" doc="Vehicle reference frame" /> -->
  <arg name="base_frame" default="lidar" doc="Vehicle reference frame" />

  <!-- gicp参数 -->
  <arg name="trans_epsilon" default="0.05" doc="The maximum difference between two consecutive transformations in order to consider convergence" />
  <arg name="cor_dist" default="0.1" doc="The newton line search maximum step length" />
  <arg name="euc_eps" default="0.05" doc="The ND voxel grid resolution" />
  <arg name="max_iterations" default="30.0" doc="The number of iterations required to calculate alignment" />
  <arg name="scan_match_score" default="0.3" doc="scan matching score" />

  <node pkg="ndt_localizer" type="gicp_localizer_node" name="gicp_localizer_node" output="screen">

    <remap from="points_raw" to="$(arg input_sensor_points_topic)" />
    <remap from="ekf_pose_with_covariance" to="$(arg input_initial_pose_topic)" />
    <remap from="pointcloud_map" to="$(arg input_map_points_topic)" />
    <remap from="ndt_pose" to="$(arg output_pose_topic)" />
    <remap from="ndt_pose_with_covariance" to="$(arg output_pose_with_covariance_topic)" />
    <remap from="diagnostics" to="$(arg output_diagnostics_topic)" />
    <param name="base_frame" value="$(arg base_frame)" />

    <param name="trans_epsilon" value="$(arg trans_epsilon)" />
    <param name="cor_dist" value="$(arg cor_dist)" />
    <param name="euc_eps" value="$(arg euc_eps)" />
    <param name="max_iterations" value="$(arg max_iterations)" />
    <param name="scan_match_score" value="$(arg scan_match_score)" />
  </node>

  <!-- urdf模型 -->
  <!-- <include file="$(find ndt_localizer)/launch/lexus.launch" /> -->

</launch>
